BLD add a mac run on Travis

jnothman opened the pull request:
Fixes #12658

jnothman committed:
BLD add a mac run on Travis

jnothman committed:
Try specifying python version for osx

jnothman committed:
Use correct miniconda installer; try generic language

jnothman committed:
No ccache on mac

jnothman committed:
Cannot deactivate if not in Python

jnothman committed:
Do not use harcoded /home/travis

jnothman committed:
More disabling ccache

qinhanmin2014 commented:
We're unable to detect it :) The kmeans mas failure is due to #12648

qinhanmin2014 committed:
try scipy 1.2.0

qinhanmin2014 committed:
scipy 1.1.0 ?

jnothman committed:
Make latest dependencies in general

jnothman commented:
Now, do we disable the fit_predict tests and add this to our regular CI?

qinhanmin2014 commented:
> Now, do we disable the fit_predict tests and add this to our regular CI?

merge #12648 to keep consistent with 0.20.X branch? I think we need this if we want to add mac CI.

qinhanmin2014 commented:
kmeans tests pass, seems that we don't need #12648 :)

jnothman commented:
So somehow they pass with latest scipy??

rth commented:
DictionaryLearning docstring fails due to a different sign in the second row of output now.

ogrisel commented on build_tools/travis/install.sh:
```diff
@@ -17,23 +17,32 @@ echo 'List files from cached directories'
     # Deactivate the travis-provided virtual environment and setup a
     # conda-based environment instead
-    deactivate
+    deactivate || :
```
What does this mean?

jnothman replied:
Apparently `:` is like pass. `deactivate` isn't applicable if we are in `language: generic`

ogrisel commented:
> So somehow they pass with latest scipy??

scipy 1.2.0 wheels seems to include openblas. scipy 1.1.0 do not (I checked the contents of the .dylibs folder of the wheel archive with unzip). Therefore scipy 1.1.0 would use the macOS accelerate system BLAS which might be the source of this non-deterministic behavior although that seems weird.

@jeremiedbb any opinion?

ogrisel commented:
> scipy 1.2.0 wheels seems to include openblas. scipy 1.1.0 do not (I checked the contents of the .dylibs folder of the wheel archive with unzip). Therefore scipy 1.1.0 would use the macOS accelerate system BLAS which might be the source of this non-deterministic behavior although that seems weird.

Actually, travis uses anaconda on the macOS build so this is unrelated. However the version of the travis build with the PCA test that fails runs openblas 0.2.20 while the latest build that does not have this failure uses openblas 0.3.3 so it might indeed be a numerical precision issue in old versions of openblas.

jeremiedbb commented:
My first though was that the test was unstable due to the way inertia is computed, which can lead to different inertia for only a permutation of the labels. But I tried with many seeds and the test seems stable. 

I Also tried with OpenBLAS 0.2.20 and OpenBLAS 0.3.3 and it's stable with both, so I does not seem to be blas instabilities either...

ogrisel committed:
Merge remote-tracking branch 'origin/master' into mac

ogrisel committed:
typo

ogrisel commented:
> DictionaryLearning docstring fails due to a different sign in the second row of output now.

The DictionaryLearning doctest was reverted in master. I merged master into this branch to check whether CI can pass now.

ogrisel committed:
joblib is required by the documentation doctests

qinhanmin2014 commented:
Is it possible to remove the 3rd Travis job?
Maybe it's better to put the mac job at the end.

ogrisel committed:
Mac build last but with MKL

Also:

- update comment to make it explicit when we test with OpenBLAS;
- remove redundant Python 3.5 linux build.

ogrisel commented:
@qinhanmin2014 I pushed a new commit with your suggestions. Also I think it's better to test the anaconda mac setup with MKL as most anaconda users on mac will probably use MKL.

ogrisel commented:
I have no idea why the previous mac build is marked failed by travis:

https://travis-ci.org/scikit-learn/scikit-learn/jobs/469979217

Here is the end of the log that shows a success:

```
[...]
==== 10538 passed, 116 skipped, 3 xfailed, 10004 warnings in 759.59 seconds ====
The command "bash build_tools/travis/test_script.sh" exited with 0.
[...]
============= 36 passed, 4 skipped, 100 warnings in 21.60 seconds ==============
The command "bash build_tools/travis/test_docs.sh" exited with 0.
0.01s$ bash build_tools/travis/test_pytest_soft_dependency.sh
The command "bash build_tools/travis/test_pytest_soft_dependency.sh" exited with 0.
cache.2
store build cache
0.01s3.62schanges detected, packing new archive
uploading PR.12824/cache-osx-1660d3122ceb6c19de16ed2ec2206da034564d90d940bc6b41e622063713f077.tgz
cache uploaded
$ source build_tools/travis/after_success.sh
```

ogrisel committed:
echo commands in after_success.sh

ogrisel committed:
Merge master

ogrisel committed:
Avoid cd into

ogrisel commented:
Thanks @jnothman. The cd command in zsh has an issue on travis under mac (there some kind of rvm-related zsh callback that fails silently). I have pushed a workaround. I will remove the `-x` now.

ogrisel committed:
Remove -x in build_tools/travis/after_success.sh

ogrisel reviewed (approved):
LGTM, once travis is green.

jnothman committed:
Comment on ||:

ogrisel commented:
Because the mac build tends to be a little bit slower than the linux build, maybe we should put it at the top of the list to maximize effective parallelization and limiting the straggler effect. WDYT?

jnothman commented:
Sounds good, @ogrisel.

ogrisel committed:
Put macOS first

ogrisel commented:
Alright the macOS build seems stable enough. Let's merge.
